storage/storage_image.go

```diff
@@ -568,7 +568,35 @@ func (s *storageImageDestination) Commit(ctx context.Context) error {
 			if err2 != nil {
 				return errors.Wrapf(err2, "error reading layer %q for blob %q", layer, blob.Digest)
 			}
-			defer diff.Close()
+			// Copy the layer diff to a file.  Diff() takes a lock that it holds
+			// until the ReadCloser that it returns is closed, and PutLayer() wants
+			// the same lock, so the diff can't just be directly streamed from one
+			// to the other.
```

Comment:
dumb image, are we going to be slowed down by this sort of buffering to disk?

Reply:
I mean, slowed down to a point we care

Reply:
Hard to say.  My guess is that computing the diff was/is the more expensive part of reading and reusing a layer's contents.  We hit this path when a layer that we have on disk matches the contents of a layer that we'd otherwise have to pull, and we can't just use the layer where it is because its inheritance chain isn't the same as the one we want.
I think that scenario becomes less and less likely as the size of the diff grows (both in the number of files and the size of their contents), and with smaller diffs, I hope the added cost of spooling the diff to disk and reading it back later is not going to be noticeable.